RIOT-OS/RIOT
# On-Chip Debugging for AVR/ATmega based boards

### Contribution description
 - Refactor/deduplicate `debug.sh`, `debug_srv.sh` and `gdb.conf` of the ATmega based boards
    - This fixes the broken `make debug` for Arduino Mega2560
- Allow to specify the hardware (debugger device) via the `DEBUGDEVICE` environment variable, instead of sticking to the device the developer of the specific board used.
    - Use the AVR Dragon by default, which is the ~~cheapest~~ least expensive programmer&debugger (thus hacker friendly) and supports the most AVR MCUs including all supported by RIOT (thus a good default for every ATmega board).
- Choose the debug protocol used based on the MCU.
    - Currently JTAG will be used unless the ATmega328P MCU is used, which supports debugWIRE instead of JTAG
- Added documentation on how to do On-Chip Debugging for the Arduino Uno
- Fixed incorrect flags for `avrdude` for the ATmega based Arduinos
     - Previously signature verification was disabled, which allows users to mistakenly flash an image intended for a completely different MCU. This is dangerous and should only be enabled by users knowing what the do for very rare and specific reasons
    - Previously auto erase was disabled. Citing the avrdude doc "it generally is a mistake to program the flash without performing an erase first" and results in corrupted flash contents when an ISP is used.
        - Added disabled of auto-erase specifically for the Arduino Mega2560 if (and only if) the default programmer is used, to work around a specific bug in the bootloader

### Testing procedure
Beware: You will need an AVR Dragon for testing. (At least for the Arduino Uno, you can test the Arduino Mega2560 with e.g. an Atmel ICE as well.)

- [ ] Arduino Uno:
    - Try to do some on-chip debugging. I'm not explaining how to do that here, so that you will test the provided doc by doing so as well
- [ ] Arduino Mega2560:
    - Try to do on-chip debugging. Read the doc (which is already in master) on how to do so. It should now work again
- [ ] Some other AVR based board ideally with a different debugger than the AVR Dragon

### Issues/PRs references
None

------

Update: Adapted description
